src/pages/index.js

```diff
@@ -82,31 +142,82 @@ const editFormValidator = new FormValidator(
 );
 const addFormValidator = new FormValidator(validationSettings, addCardModal);
 
+const editAvatarValidator = new FormValidator(
+  validationSettings,
+  avatarEditModal
+);
+
 //Initialize all instances
-cardSection.renderItems(initialCards);
+api.getInitialCards().then((cards) => {
+  cardSection.renderItems(cards);
+});
+
 cardPreviewPopup.setEventListeners();
 editPopup.setEventListeners();
 addPopup.setEventListeners();
+editAvatarPopup.setEventListeners();
 editFormValidator.enableValidation();
 addFormValidator.enableValidation();
+editAvatarValidator.enableValidation();
 
 /*--------------------------------------------------*/
 /*                      Functions                   */
 /*--------------------------------------------------*/
 
+function handleDeleteCardClick() {
+  confirmDeleteModal.open();
+  confirmDeleteModal.setSubmitAction(() => {
+    const id = this.getId();
+
+    api
+      .removeCard(id)
+      .then((res) => {
+        this.handleDeleteCard();
+      })
+      .catch((err) => {
+        console.error(err);
+      });
+  });
+}
+
+function handleCardLikeClick(cardId, isLiked) {
+  if (isLiked) {
+    api
+      .removeCardLike(cardId)
+      .then((cardData) => this.setLike(cardData.isLiked));
+  } else {
+    api.addCardLike(cardId).then((cardData) => {
+      this.setLike(cardData.isLiked);
+    });
+  }
+}
+
 function handleProfileEditSubmit(obj) {
-  const { title, description } = obj;
-  userInfo.setUserInfo(title, description);
-  editPopup.close();
+  editPopup.renderLoading(true, "Saving...");
+  api
+    .updateProfileInfo(obj)
+    .then((res) => {
+      const { name, about } = res;
+      userInfo.setUserInfo(name, about);
+      editPopup.close();
+    })
+    // console.log(obj)
+    .finally(() => {
+      editPopup.renderLoading(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```